feat: add parameter description (#271)
autowarefoundation/autoware-documentation: docs/contributing/coding-guidelines/ros-nodes/parameters.md

# Parameters

> ## ROS parameters

Author: Resolved: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/fa5b7c7b27b9930c32d815e3464faa802ab6e176

> ## ROS parameters

Reviewer: @kminoda I'm sorry, this was not resolved.

```suggestion
```

> ## ROS parameters

Author: Right, thank you! @kenji-miyake 
Resolved: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/96a8a931649e9ef7d7f0167cc27db3561519d21b

> The ROS packages in Autoware may have ROS parameter(s) for setting some variables that should be customized depending on the applications.

Author: Resolved: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/fa5b7c7b27b9930c32d815e3464faa802ab6e176

> It is recommended not to set default values when you declare ROS parameters to avoid unintended behaviors.

Author: You mean about a line `It is recommended not to set default values when you declare ROS parameters to avoid unintended behaviors.`, right?

I could not find any part that refers to "avoid using default parameters" in ROS 2 coding guidelines, so I feel it's better to leave it just as a recommendation for now.
(Or maybe we can just remove this line, since I cannot correctly recall if I had enough discussion about this recommentation...)

> It is recommended not to set default values when you declare ROS parameters to avoid unintended behaviors.

Author: @kenji-miyake 
> It's an Autoware's guideline, proposed by TIER IV.

I see, I didn't know that

> It's also okay for me. It's better to create a discussion thread and investigate more.

OK, I also agree that rules can be determined later after some discussions on AWF. So I will leave the PR as it is for now.

> It is recommended not to set default values when you declare ROS parameters to avoid unintended behaviors.

Reviewer: @kminoda How about this?

```suggestion
It is recommended not to set default values when declaring ROS parameters to avoid unintended behaviors due to accidental use of default values.
Instead, set parameters from configuration files named `*.param.yaml`.
```

> It is recommended not to set default values when you declare ROS parameters to avoid unintended behaviors.

Author: Thank you! I take that idea.
Resolved: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/843889d8e7a65d7ee4d5cd42633625a86f9948aa

> ## ROS Parameter files

Author: Indirectly resolved: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/fa5b7c7b27b9930c32d815e3464faa802ab6e176

Autoware has the following two types of parameter files for ROS packages:

- Node parameter file: the reference parameter for the node package

>   - e.g., [the parameter for the `behavior_path_planner` package](https://github.com/autowarefoundation/autoware.universe/tree/main/planning/behavior_path_planner/config)

Author: Thank you! What do you mean by permanent link? Do you mean such as using a link of a specific commit?

>   - e.g., [the parameter for the `behavior_path_planner` package](https://github.com/autowarefoundation/autoware.universe/tree/main/planning/behavior_path_planner/config)

Reviewer: @kminoda Yes, open https://github.com/autowarefoundation/autoware.universe/tree/main/planning/behavior_path_planner/config and press `Y`.
Then you'll get the fixed link.
https://github.com/autowarefoundation/autoware.universe/tree/030590645d835127f49d1cbd0715d9966ad5df11/planning/behavior_path_planner/config

>   - e.g., [the parameter for the `behavior_path_planner` package](https://github.com/autowarefoundation/autoware.universe/tree/main/planning/behavior_path_planner/config)

Author: Resolved: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/fa5b7c7b27b9930c32d815e3464faa802ab6e176

>   - e.g., [the parameter for the `behavior_path_planner` package](https://github.com/autowarefoundation/autoware.universe/tree/main/planning/behavior_path_planner/config)

Author: (I didn't know that trick, thank you)
Resolved: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/c59c6270f85943966f5a3156a8ae22bd472bb20b

> - Launch parameter file: [the reference parameter for the `autoware_launch` package](https://github.com/autowarefoundation/autoware_launch/tree/main/autoware_launch/config)

Author: Right, resolved here: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/7ccad2e03b88789e0de6291b93564842cd8a2f42

All the parameter files should have the `.param.yaml` suffix so that the auto-format can be applied properly.

### Node parameter

> Node parameter files store the default parameters provided for each package in Autoware Universe.

Author: Resolved: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/fa5b7c7b27b9930c32d815e3464faa802ab6e176

> All the nodes in Autoware **must** have the node parameter file.

Reviewer: [Question] For what reason?

Author: I thought it is mandatory so that all the nodes should be able to be tuned depending on the vehicles. But then I realized this is not always the case (some node may have no parameters, and we don't need parameter files in such case).

Is this also what your question intended? Then, I will fix this line.

Reviewer: Yes. I wanted to know the background reason.
Let's write the rationales and exceptions.

Author: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/7187ce7bd0d4439a716fb1842ddd476dfb7cc25e
How about this?

Reviewer: Cool!

[nits]
- no comma?
- the node have -> the node has?

Author: As well as resolving the grammar issue, I also revised the sentence https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/f420a284eca97bd8a5d203cb85d9af67859d1d8f

Reviewer: :+1:

For `FOO_package`, the parameter is expected to be stored in `FOO_package/config`.

> Also, the launch file for individual packages must load node parameter by default.

Author: Resolved: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/fa5b7c7b27b9930c32d815e3464faa802ab6e176

> ### Launch parameter

Author: Resolved:

> Launch parameter files store the parameters that are used when launching Autoware, which are stored under `autoware_launch` repository. Basically, launch parameters are the subset of node parameters. It is expected that the users tune launch parameters in `autoware_launch`, and thus, a node parameter file should also be added to launch parameter if the parameter may be tuned depending on the vehicle.

Reviewer: Could you simplify this sentence? I feel it's a little difficult.

Author: How about this? https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/7ccad2e03b88789e0de6291b93564842cd8a2f42

Reviewer: It's simple enough!

# Parameters

The ROS packages in Autoware have ROS parameters. You need to customize the parameters depending on your applications.
It is recommended not to set default values when declaring ROS parameters to avoid unintended behaviors due to accidental use of default values.

> Instead, set parameters from configuration files named `*.param.yaml`.
> 

Reviewer 2: ```suggestion
Instead, set parameters from configuration files named `*.param.yaml`.

Also check out the official documentation for [understanding ROS 2 parameters](https://docs.ros.org/en/rolling/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Parameters/Understanding-ROS2-Parameters.html).

```

Maybe a small line to link to the official docs.

Reviewer: Considering https://developers.google.com/style/sentence-structure, how about this?

```suggestion
Instead, set parameters from configuration files named `*.param.yaml`.

For understanding ROS 2 parameters, also check out the official documentation [Understanding parameters](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Parameters/Understanding-ROS2-Parameters.html).
```

Author: @xmfcx @kenji-miyake 
Thank you! Reflected: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/70f9b6c8d723c92a5f34db9bf20550db72775a21

It is recommended not to set default values when you declare ROS parameters to avoid unintended behaviors.

## Parameter files

Autoware has the following two types of parameter files for ROS packages:

- Node parameter file: the reference parameter for the node package
  - For example, [the parameter for the `behavior_path_planner` package](https://github.com/autowarefoundation/autoware.universe/tree/main/planning/behavior_path_planner/config)
- Launch parameter file: [the reference parameter for the `autoware_launch` package](https://github.com/autowarefoundation/autoware_launch/tree/main/autoware_launch/config)

All the parameter files should have the `.param.yaml` suffix so that the auto-format can be applied properly.

### Node parameter

Node parameter files store the default parameters provided for each package in Autoware.
All the nodes in Autoware **must** have the node parameter file.
For `FOO_package`, the parameter is expected to be stored in `FOO_package/config`.
Also, the launch file for individual packages must load node parameter by default:

```xml
<launch>
  <arg name="foo_node_param_path" default="$(find-pkg-share FOO_package)/config/foo_node.param.yaml" />

>   <node ...>

Reviewer: ```suggestion
  <node pkg="FOO_package" exec="foo_node">
```

Author: Resolved: https://github.com/autowarefoundation/autoware-documentation/pull/271/commits/ce3267969085617fe5d77a9bbfe79c7c0cc632fd
